File: pixleesdk/src/main/java/com/pixlee/pixleesdk/data/api/BasicAPI.java
package com.pixlee.pixleesdk.data.api;

import com.pixlee.pixleesdk.PXLPhoto;
import com.pixlee.pixleesdk.data.MediaResult;
import com.pixlee.pixleesdk.data.PhotoResult;
import com.serjltt.moshi.adapters.Wrapped;

import java.util.List;

import okhttp3.MultipartBody;
import okhttp3.RequestBody;
import retrofit2.Call;
import retrofit2.http.Body;
import retrofit2.http.GET;
import retrofit2.http.Header;
import retrofit2.http.Multipart;
import retrofit2.http.POST;
import retrofit2.http.Part;
import retrofit2.http.Path;
import retrofit2.http.Query;

/**
 * Restful API Access Object for basic APIs
 * Retrofit Document: https://square.github.io/retrofit/
 */
public interface BasicAPI {
    @GET("albums/from_sku")
    Call<PhotoResult> getPhotosWithSKU(
            @Query("sku")String sku,
            @Query("api_key")String api_key,
            @Query("filters")String filters,
            @Query("sort")String sort,
            @Query("per_page")int per_page,
            @Query("page")int page
    );

    @GET("albums/{album_id}/photos")
    Call<PhotoResult> getPhotosWithID(
            @Path("album_id")String album_id,
            @Query("api_key")String api_key,
            @Query("filters")String filters,
            @Query("sort")String sort,
            @Query("per_page")int per_page,
            @Query("page")int page
    );

    @GET("media/{album_photo_id}")
    @Wrapped(path = {"data"})
    Call<PXLPhoto> getMedia(@Path("album_photo_id")String album_photo_id, @Query("api_key")String api_key);

    @POST("media")
    Call<MediaResult> postMedia(
            @Header("Signature") String Signature,
            @Query("api_key")String api_key,
            @Body RequestBody body
    );

    @Multipart
    @POST("media/file")
    Call<MediaResult> uploadImage(
            @Header("Signature") String Signature,
            @Query("api_key")String api_key,
            @Part List<MultipartBody.Part> partList);
}

File: pixleesdk/src/main/java/com/pixlee/pixleesdk/data/api/BasicAPI.kt
package com.pixlee.pixleesdk.data.api

import com.pixlee.pixleesdk.data.MediaResult
import com.pixlee.pixleesdk.data.PXLPhoto
import com.pixlee.pixleesdk.data.PhotoResult
import com.serjltt.moshi.adapters.Wrapped
import okhttp3.MultipartBody
import okhttp3.RequestBody
import retrofit2.Call
import retrofit2.http.*

/**
 * Restful API Access Object for basic APIs
 * Retrofit Document: https://square.github.io/retrofit/
 */
interface BasicAPI {
    @GET("api/v2/albums/from_sku")
    fun getPhotosWithSKU(
            @Query("sku") sku: String,
            @Query("api_key") api_key: String,
            @Query("filters") filters: String?,
            @Query("sort") sort: String?,
            @Query("per_page") per_page: Int,
            @Query("page") page: Int,
            @Query("region_id") region_id: Int?

    ): Call<PhotoResult>

    @GET("api/v2/albums/{album_id}/photos")
    fun getPhotosWithID(
            @Path("album_id") album_id: String,
            @Query("api_key") api_key: String,
            @Query("filters") filters: String?,
            @Query("sort") sort: String?,
            @Query("per_page") per_page: Int,
            @Query("page") page: Int,
            @Query("region_id") region_id: Int?
    ): Call<PhotoResult>

    @GET("api/v2/media/{album_photo_id}")
    @Wrapped(path = ["data"])
    fun getMedia(@Path("album_photo_id") album_photo_id: String, @Query("api_key") api_key: String): Call<PXLPhoto>

    @POST("api/v2/media")
    fun postMediaWithURI(
            @Header("Signature") Signature: String,
            @Query("api_key") api_key: String?,
            @Body body: RequestBody
    ): Call<MediaResult>

    @Multipart
    @POST("api/v2/media/file")
    fun postMediaWithFile(
            @Header("Signature") Signature: String,
            @Query("api_key") api_key: String?,
            @Part partList: List<MultipartBody.Part>): Call<MediaResult>
}
